Initial commit for obsolete and deprecated features documentation

rkutteh opened the pull request:
# CABLE

## Description

I have created four files to start documenting deprecated and obsolete features of the code for both users and developers. I have included some text to get things started. I will be adding content to all these files as I go through my second iteration of the CABLE code. Please do not remove this branch for the foreseeable future.  Thanks.

You can link issues by using a supported keyword in the pull request's description or in a commit message:

Fixes #(issue)

## Type of change

Please delete options that are not relevant.

- [X ] New or updated documentation

## Checklist

- [ ] The new content is accessible and located in the appropriate section.
- [ ] I have checked that links are valid and point to the intended content.
- [ ] I have checked my code/text and corrected any misspellings

Please add a reviewer when ready for review.

tammasloughran commented:
I think some items for these sections need to be added to the [`mkdocs.yml`](https://github.com/CABLE-LSM/CABLE/blob/9e7da02f8f3f29cdec73a5b8d2a9f708912dceb7/documentation/mkdocs.yml#L94) file.

ccarouge reviewed (commented):
Tammas is right, you need to add these files to the navigation in `mkdocs.yml`. Let me know if you prefer me to do it.

What is the reason for splitting between User and Developer features? How are we going to decide if a feature should be in one or the other? I would imagine the User features would be a subset of the Developer features so do we copy all the User features to the Developer features file?

rkutteh commented:
@ccarouge Just to answer your three questions, as I have probably argued in some previous meetings, in my view developers and users are two entirely independent non-intersecting sets of people (i.e., neither is a subset of the other). For example, I am currently working on the GW part of the code but unlikely to ever do serious science using CABLE, whereas **most** scientist who use CABLE to do science are unlikely to be involved in its code development.  

- **What is the reason for splitting between User and Developer features?**  The user features concern only the user interface with the code (I/O, settings,...) and a user should not be expected to browse through a file containing many developer features when checking if an interface feature has been deprecated or removed altogether. The only thing the user cares about is the user interface settings and not any internal variables, routines ... in the code. So that is the first reason for splitting between User and Developer features.

- **I would imagine the User features would be a subset of the Developer features so do we copy all the User features to the Developer features file?** The user is interested in the interface/user features only, the developer is interested mainly in the internal code which is a black box to the user and occasionally in the interface code. Some developers may be working on part of the code entirely unrelated to user features while other developers may have some interaction with user features. So my expectation is that a user would only need to look inside the user features file, whereas a developer would need to look inside the developer features file primarily and only occasionally inside the user features file when concerned with I/O issues for example. I added links inside each file to the other with this strategy in mind.

- **How are we going to decide if a feature should be in one or the other?** Easy, if a user will never have a chance to interact with a feature then it is a developer feature and the user need not be bothered when it becomes obsolete or deprecated as it will not affect the user interface at all. For example, we may decide to remove a redundant subroutine or chunk of the code, rename some internal variables, split the driver into several files,...the user need not worry about this stuff as the code will still behave the same to them. So all deprecated/obsolete interface features go into the user file and all other deprecated/obsolete features go into the developer file.

Let me know what you think and then I can do the navigation issue.

ccarouge commented on documentation/docs/developer_guide/obsolete_and_deprecated_features/deprecated_features.md:
```diff
@@ -0,0 +1,36 @@
+
+## List of deprecated developer features ##
+
+-------------------------------------------------------------------------------
+
+**Deprecated developer feature:** enter the exact syntax if appropriate
+
+**Deprecated from CABLE version:** enter version number and date
+
+**Replacement developer feature?:** describe the current alternative feature
+
+**Comments:** e.g., describe the locations of the deprecated and replacement features and their purposes
+
+-------------------------------------------------------------------------------
+
+**Deprecated developer feature:** enter the exact syntax if appropriate
+
+**Deprecated from CABLE version:** enter version number and date
+
+**Replacement developer feature?:** describe the current alternative feature
+
+**Comments:** e.g., describe the locations of the deprecated and replacement features and their purposes
+
+-------------------------------------------------------------------------------
```
I would prefer the following format:

```
| Feature name | <div style="width:150px"> Reason for deprecation </div> | Replacement? | CABLE version | Pull Request |
| ------------ | ------------------------------------------------------- | ------------ | ------------- | ------------ |
| enter the exact syntax of the feature | short reason for deprecation | add the current alternative if any | enter the version number | [#enter the pull request number](copy PR's URL) |
| enter the exact syntax of the feature | short reason for deprecation | add the current alternative if any | enter the version number | [#enter the pull request number](copy PR's URL) |
```

I know you are worried about having a very long table. But otherwise, we will have an even longer page.

rkutteh replied:
I have adopted your format

ccarouge commented on documentation/docs/developer_guide/obsolete_and_deprecated_features/obsolete_features.md:
```diff
@@ -0,0 +1,37 @@
+# Obsolete features for CABLE developers #
```
I would merge the pages for deprecated and obsolete together. Then use the table of contents on each page to navigate between the 2.

rkutteh replied:
I have merged deprecated and obsolete together.

ccarouge commented on documentation/docs/developer_guide/obsolete_and_deprecated_features/deprecated_features.md:
```diff
@@ -0,0 +1,36 @@
+# Deprecated features for CABLE developers #
+
+Coding with the deprecated developer features described here is discouraged as they are scheduled to become obsolete
+and removed from future versions of the CABLE code. 
```
We could add:

Development that includes any deprecated or obsolete feature will require modifications before being accepted for a CABLE release.

rkutteh replied:
Ok I added this bit

ccarouge commented on documentation/docs/user_guide/obsolete_and_deprecated_features/deprecated_features.md:
```diff
@@ -0,0 +1,37 @@
+# Deprecated features for CABLE users #
```
Same comment about merging the pages for deprecated and obsolete features.

rkutteh replied:
I have merged deprecated and obsolete together.

ccarouge commented on documentation/docs/user_guide/obsolete_and_deprecated_features/deprecated_features.md:
```diff
@@ -0,0 +1,37 @@
+  
+The template below can be used to report deprecated user features (one feature per block):
+
+## List of deprecated user features ##
```
Same comment about the format to list a feature.

rkutteh replied:
I have adopted your format

ccarouge reviewed (changes requested):
Let me know what you think about my suggestions.

Also, remember to add the pages to the site navigation in `mkdoc.yml`, I am happy to help with that if you prefer.

rkutteh commented:
@ccarouge I have made all the changes you requested. I also cleaned up the text inside the tables and made the change to mkdocs.yml as suggested by @tammasloughran. Please note that although I merged the two files into one as requested, I left each resulting file in a directory by itself for now, as we might end up needing to add more documentation in that directory in the future.

ccarouge commented:
@rkutteh ready for you to merge your branch in.
